File: client/src/lib/i18n.test.ts
import assert from 'node:assert/strict';
import { describe, it } from 'node:test';
import { TRANSLATIONS, getTranslation, getPlural, type LanguageCode } from './i18n.dictionary.ts';

describe('Web Client i18n Translations', () => {
  const languages: LanguageCode[] = ['en', 'es', 'fr', 'de', 'pt', 'it'];

  it('interpolates {vars} and picks plural forms', () => {
    assert.strictEqual(getTranslation('n_ready', 'en', { n: 4 }), '4 ready');
    assert.strictEqual(getTranslation('n_ready', 'es', { n: 4 }), '4 listas');
    assert.strictEqual(getPlural('streak_days', 1, 'en'), '1 day');
    assert.strictEqual(getPlural('streak_days', 3, 'de'), '3 Tage');
    // Unknown placeholder is left intact rather than becoming "undefined".
    assert.strictEqual(getTranslation('word_around', 'en', {}), 'Word around {time}');
  });
  const baseKeys = Object.keys(TRANSLATIONS.en);

  languages.forEach((lang) => {
    it(`has complete translation dictionary for ${lang}`, () => {
      const dict = TRANSLATIONS[lang];
      assert.ok(dict);

      baseKeys.forEach((key) => {
        assert.ok(dict[key]);
        assert.strictEqual(typeof dict[key], 'string');
        assert.ok(dict[key].length > 0);
      });
    });
  });

  it('translates correctly with fallback to English for unknown keys', () => {
    assert.strictEqual(getTranslation('nav_discover', 'es'), 'Descubrir');
    assert.strictEqual(getTranslation('nav_discover', 'fr'), 'Découvrir');
    assert.strictEqual(getTranslation('nav_discover', 'de'), 'Entdecken');
    assert.strictEqual(getTranslation('unknown_key_test', 'es'), 'unknown_key_test');
  });
});

File: mobile/test/i18n_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:harmonix_mobile/utils/i18n.dart';

void main() {
  const langs = ['en', 'es', 'fr', 'de', 'pt', 'it'];

  test('every language has every key of the base and extra dictionaries', () {
    final baseKeys = AppI18n.translations['en']!.keys.toSet();
    final extraKeys = AppI18n.extra['en']!.keys.toSet();
    for (final lang in langs) {
      final base = AppI18n.translations[lang]!;
      final extra = AppI18n.extra[lang]!;
      expect(base.keys.toSet(), baseKeys, reason: 'base keys for $lang');
      expect(extra.keys.toSet(), extraKeys, reason: 'extra keys for $lang');
      for (final v in [...base.values, ...extra.values]) {
        expect(v.trim(), isNotEmpty);
      }
    }
  });

  test('interpolates {vars}, falls back to English, and picks plurals', () {
    expect(AppI18n.t('n_ready', 'es', {'n': 4}), '4 listas');
    expect(AppI18n.t('n_ready', 'xx', {'n': 4}), '4 ready');
    expect(AppI18n.t('unknown_key'), 'unknown_key');
    expect(AppI18n.plural('streak_days', 1, 'en'), '1 day');
    expect(AppI18n.plural('streak_days', 3, 'de'), '3 Tage');
    expect(AppI18n.plural('to_review', 2, 'pt'), '2 palavras para revisar →');
  });
}
